RecoEgamma/EgammaPhotonProducers/src/GEDPhotonGSCrysFixer.cc

```diff
@@ -0,0 +1,257 @@
+#include <vector>
+#include <memory>
+
+#include "RecoEgamma/EgammaPhotonProducers/interface/GEDPhotonGSCrysFixer.h"
+
+#include "FWCore/Utilities/interface/isFinite.h"
+
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+#include "Geometry/CaloEventSetup/interface/CaloTopologyRecord.h"
+
+#include "DataFormats/VertexReco/interface/Vertex.h"
+#include "DataFormats/EgammaCandidates/interface/PhotonCore.h"
+#include "DataFormats/EgammaCandidates/interface/Photon.h"
+
+#include "RecoEcal/EgammaCoreTools/interface/EcalClusterFunctionBaseClass.h" 
+#include "RecoEcal/EgammaCoreTools/interface/EcalClusterFunctionFactory.h" 
+#include "RecoEgamma/EgammaTools/interface/GainSwitchTools.h" 
+
+typedef edm::ValueMap<reco::PhotonRef> PhotonRefMap;
+
+namespace {
+  reco::PhotonCoreRef
+  getNewCore(reco::Photon const& oldPho, edm::Handle<reco::PhotonCoreCollection> const& newCores, edm::ValueMap<reco::SuperClusterRef> const& scMap)
+  {
+    for (unsigned iC(0); iC != newCores->size(); ++iC) {
+      reco::PhotonCoreRef newCore(newCores, iC);
+      auto&& oldSCRef(scMap[newCore]);
+      if (oldSCRef.isNonnull() && oldSCRef == oldPho.superCluster())
+	return newCore;
+    }
+    return reco::PhotonCoreRef(newCores.id());
+  }
+}
+
+
+GEDPhotonGSCrysFixer::GEDPhotonGSCrysFixer(const edm::ParameterSet& config) :
+  energyCorrector_(config, consumesCollector())
+{
+  getToken(inputPhotonsToken_, config, "photons");
+  getToken(newCoresToken_, config, "newCores");
+  getToken(newCoresToOldSCMapToken_, config, "newCores");
+  getToken(ebHitsToken_, config, "barrelEcalHits");
+  getToken(verticesToken_, config, "primaryVertexProducer");
+
+  if(config.existsAs<edm::ParameterSet>("regressionConfig")) {
+    auto&& collector(consumesCollector());
+    energyCorrector_.gedRegression()->setConsumes(collector);
+  }
+
+  // output photon collection
+  produces<reco::PhotonCollection>();
+  // new photon -> old photon
+  produces<PhotonRefMap>();
+}
+
+GEDPhotonGSCrysFixer::~GEDPhotonGSCrysFixer() 
+{
+}
+
+void
+GEDPhotonGSCrysFixer::beginRun(edm::Run const&, edm::EventSetup const& _setup)
+{
+  energyCorrector_.init(_setup);
+
+  edm::ESHandle<CaloGeometry> geometryHandle;
+  _setup.get<CaloGeometryRecord>().get(geometryHandle);
```

Comment:
this better be moved to beginLumi

we've seen the pointers move in the past on IOV boundaries inside a run